feat: add alert rule options to not send alert on state recovery and to send regardless of state change

mmdoogie opened the pull request:
This was previous requested in #3723 that went stale.  I recently had a use case for it and decided to add it.

Added a checkbox to the alert rule builder that's used to determine whether or not to include `noRecoveries` in the TICKscript.
I also saw that `stateChangesOnly` was already in the structure so I exposed that in a similar manner.

![image](https://user-images.githubusercontent.com/16321466/158329842-97df62a7-4016-4614-927d-b79ab140bced.png)

Additionally, the swagger.json didn't look like it matches the current response to the API even before the changes, it was fixed with a new flag added.


  - [x] CHANGELOG.md updated with a link to the PR (not the Issue)
  - [x] [Well-formatted commit messages](https://www.conventionalcommits.org/en/v1.0.0-beta.3/)
  - [x] Rebased/mergeable
  - [x] Tests pass
  - [x] swagger.json updated (if modified Go structs or API)
  - [x] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)

sranka commented:
Thank you @mmdoogie for your contribution. Can you please change your commit messages to [semantic format](https://seesparkbox.com/foundry/semantic_commit_messages).

mmdoogie commented:
Updated messages as requested.  Also synced up the swagger.json with the existing state of alerts/alertNode from the codebase and added the new object parameter.

sranka commented on ui/src/kapacitor/reducers/rules.js:
```diff
@@ -16,7 +16,7 @@ export default function rules(state = {}, action) {
           values: defaultRuleConfigs.threshold,
           message: '',
-          alertNodes: {},
+          alertNodes: {stateChangesOnly: true},
```
yes, `stateChangesOnly` seems to be the default

sranka commented on ui/src/kapacitor/components/RuleHandlers.tsx:
```diff
@@ -132,6 +134,33 @@ class RuleHandlers extends PureComponent<Props, State> {
       <div className="rule-section">
         <h3 className="rule-section--heading">Alert Handlers</h3>
         <div className="rule-section--body">
+          <div className="rule-section--row rule-section--row-first rule-section--border-bottom">
+            <p>Alert Options:</p>
+            <div className="form-control-static">
+              <input
+                name="noRecoveries"
+                id="noRecoveries"
+                type="checkbox"
+                defaultChecked={rule.alertNodes.noRecoveries}
+                onClick={this.handleNoRecoveries}
+              />
+              <label htmlFor="noRecoveries">
+                Don't send alert on condition recovery
+              </label>
+            </div>
+            <div className="form-control-static">
+              <input
+                name="stateChangesOnly"
+                id="stateChangesOnly"
+                type="checkbox"
+                defaultChecked={rule.alertNodes.stateChangesOnly}
+                onClick={this.handleStateChangesOnly}
+              />
+              <label htmlFor="stateChangesOnly">
+                Send alert only when condition state changes
+              </label>
+            </div>
+          </div>
```
This change looks like this:

![image](https://user-images.githubusercontent.com/16321466/153740343-fba95eff-fffb-4e01-ada2-5b6ea094890c.png)

which might make users a bit confused because:
- the new options appear first so that one might think that it is the most important thing of this section ... it is not, the options must be the last 
- the options appear fused, there should be some space to separate them
- the options appear even when no rule are selected

It will look better with these disruptions fixed:

![image](https://user-images.githubusercontent.com/16321466/153740586-2c2cff95-dff1-4fab-86b5-f1f56747f606.png)

I propose to have the new options under `{mappedHandlersOnThisAlert.length ? ` :
```jsx
          {mappedHandlersOnThisAlert.length ? (
            <>
              <div className="rule-message--endpoints rule-section--border-bottom">
                <HandlerTabs
                  handlersOnThisAlert={mappedHandlersOnThisAlert}
                  selectedHandler={selectedHandlerWithText}
                  handleChooseHandler={this.handleChooseHandler}
                  handleRemoveHandler={this.handleRemoveHandler}
                />
                <HandlerOptions
                  selectedHandler={selectedHandler}
                  handleModifyHandler={this.handleModifyHandler}
                  updateDetails={ruleActions.updateDetails}
                  rule={rule}
                  onGoToConfig={onGoToConfig}
                  onTestHandler={onTestHandler}
                  validationError={validationError}
                />
              </div>
              <div className="rule-section--row rule-section--row-last">
                <div className="form-control-static">
                  <input
                    name="stateChangesOnly"
                    id="stateChangesOnly"
                    type="checkbox"
                    defaultChecked={rule.alertNodes.stateChangesOnly}
                    onClick={this.handleStateChangesOnly}
                  />
                  <label htmlFor="stateChangesOnly">
                    Send alert only when condition state changes
                  </label>
                </div>
                <div className="form-control-static" style={{marginLeft: 5}}>
                  <input
                    name="noRecoveries"
                    id="noRecoveries"
                    type="checkbox"
                    defaultChecked={rule.alertNodes.noRecoveries}
                    onClick={this.handleNoRecoveries}
                  />
                  <label htmlFor="noRecoveries">
                    Don't send alert on condition recovery
                  </label>
                </div>
              </div>
            </>
          ) : null}
```

sranka reviewed (changes requested):
Thank you for updating the swagger file to be closer to reality. It still needs some polishing to be approved.  
- every new alert must be created with a generated ID
- `kapacitor/ast_test.go` needs to test the new `noRecoveries` property
- UI must be improved

sranka commented:
@mmdoogie please let me know if you need help with the changes.

mmdoogie commented:
Sorry it has taken me a moment to get back to this.  UI rearranged and better spaced now, definitely better:
![2](https://user-images.githubusercontent.com/24367033/158038929-f453266b-1fd5-41ef-9b49-11c8440907a8.png)
also not present unless a handler has been added:
![1](https://user-images.githubusercontent.com/24367033/158038957-9da5f460-a2c4-4ad9-a79a-ee5d5d4dcfd9.png)

Will add a test for the new parameters and fix the ID part soon.

mmdoogie commented:
There's an interaction with the pipeline/tick/AST code in kapacitor -- when a new rule is created and POSTed to chronograf, the kapacitor/alerts.go addAlertNodes uses that functionality to build the alert nodes tickscript.  It appears to be parsing the flags correctly up to that point but the build/format process only returns the actual alert nodes and not any of the flag parameters.

sranka commented:
> There's an interaction with the pipeline/tick/AST code in kapacitor -- when a new rule is created and POSTed to chronograf, the kapacitor/alerts.go addAlertNodes uses that functionality to build the alert nodes tickscript. It appears to be parsing the flags correctly up to that point but the build/format process only returns the actual alert nodes and not any of the flag parameters.

This was a UI issue that you already fixed in [d806ba8](https://github.com/influxdata/chronograf/pull/5856/commits/d806ba8f5f371a58402968bc5b2fb909d37d51f9), and I also modified a test to cover it in [592d2d5](https://github.com/influxdata/chronograf/pull/5856/commits/592d2d5399a24c295b53c213729baa8c5e3351b2)

sranka reviewed (approved):
Thank you @mmdoogie for addressing my feedback. I rebased your changes on top of the latest master and also tested the new functionality manually to see it in action. It works fine. I also added an extra test to verify that the flags are kept when updating alert nodes.
